iam_cert return arn and allow use with ansible vault

wrouesnel opened the pull request:
##### SUMMARY
Return certificate ARN in all success cases with `iam_cert`, allowing idempotent provisioning of other AWS services like elastic load balancers directly in the same playbook (which require an ARN for SSL for example).

There is also an included fix to allow duplicates to be correctly detected when the certificate body includes the authority chain directly (the data returned by AWS previously did not match).

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
iam_cert

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 4339be2fa8) last updated 2017/06/18 00:39:45 (GMT +1100)
  config file = None
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible-work/ansible/lib/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```
##### ADDITIONAL INFORMATION

wrouesnel commented:
Migrated from https://github.com/ansible/ansible-modules-core/pull/2219

wrouesnel commented:
What's needed to move this along? This has been pending for >1 year.

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -269,10 +276,21 @@ def main():
-        key = open(module.params.get('key'), 'r').read().rstrip()
+        try:
+            cert = open(module.params.get('cert'), 'r').read().rstrip()
+        except IOError:
```
I love the idea of having input from a file or string for this module, although I'm curious if this best method or if we should have 2 separate vars similar to copy etc.  eg.  mutually exclusive vars cert_path and cert_content

I too would like to see this implemented soon along with [my PR for iam_cert](https://github.com/ansible/ansible/pull/20622)

wrouesnel replied:
I think at the time I was copying another module which did it this way, but I can't remember what. Standards of coding in ansible have migrated since I first wrote this.

ProjectRice replied:
I don't have the answer, but was hoping to prompt discussion about it and maybe get this moved along in the process.

ProjectRice replied:
Could this pattern cause confusion in a check mode scenario?

s-hertel replied:
This module doesn't support check mode. I'm not sure if this is the best method to do this either though.

s-hertel replied:
I'm going to add a commit to do `if os.path.exists(module.params.get('cert')):` instead of try/except.

s-hertel commented:
shipit

This looks good to me. I'm sorry @wrouesnel that this hasn't had much attention sooner- it looks like in the old repo there was some enthusiasm for this. I broke the descriptions into two lines since shippable balks at lines longer than 160 characters now. I'll make a follow up PR for preexisting pep8 issues with this module, maybe after ProjectRice's pr is implemented.

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -279,20 +280,20 @@ def main():
     dup_ok = module.params.get('dup_ok')
     if state == 'present':
-        try:
+        if os.path.exists(module.params.get('cert')):
```
Much nicer :+1: 

Standards wise do we care that these are defined as paths?

s-hertel replied:
Oh, good catch. I guess `os.path.isfile` probably makes more sense?

ProjectRice replied:
My understanding is exists will return true for a folder, isfile will not.  So probably yes. isfile will be the best fit.

s-hertel commented:
re shipit

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -280,18 +280,18 @@ def main():
     cert_chain = module.params.get('cert_chain')
     dup_ok = module.params.get('dup_ok')
     if state == 'present':
```
I"m probably making it harder to merge my PR so this might be best left for later update, however we should we take the opportunity to improve fix it properly.

Assigning local vars for cert and key since they now get read several times from modules.params

Only checking if the file exists or trying to read the string if the values exist.  Otherwise rstrip will fail on a None.

```
cert = module.params.get('cert')
if cert:
    if os.path.isfile(cert):
       cert = open(cert, 'r').read().rstrip()
    else:
        cert = unicode(cert.rstrip())

```

ProjectRice replied:
stupid indents

ryansb commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -269,10 +280,23 @@ def main():
+            if os.path.isfile(module.params.get('key')):
+                key = open(module.params.get('key'), 'r').read().rstrip()
+            else:
+                key = unicode(module.params.get('key').rstrip())
```
unicode won't always work (Python 2/3 split) so please use the to_text from ansible's utils instead.

dschneller commented:
This is exactly what I am looking for. Is there anything left to be done, apart from the review? Any chance of getting this merged into devel soon?

ProjectRice commented:
Any more movement on this one?  I'm still keen to get this and https://github.com/ansible/ansible/pull/20622 in soon.  Probably no point refactoring https://github.com/ansible/ansible/pull/20622 yet though as it will only need to be done again after this is merged.

ProjectRice commented on lib/ansible/modules/cloud/amazon/iam_cert.py:
```diff
@@ -269,10 +281,23 @@ def main():
     cert_chain = module.params.get('cert_chain')
     dup_ok = module.params.get('dup_ok')
     if state == 'present':
-        cert = open(module.params.get('cert'), 'r').read().rstrip()
```
Just noticed....   These assignments need to be here (but modified) otherwise the following if statements will fail.
I would also move the cert_chain assignment into if state == 'present' block.

So along these lines....
```
    state = module.params.get('state')
    name = module.params.get('name')
    path = module.params.get('path')
    new_name = module.params.get('new_name')
    new_path = module.params.get('new_path')   
    dup_ok = module.params.get('dup_ok')

    if state == 'present':
        cert = module.params.get('cert')
        if cert is not None:
            if os.path.isfile(cert):
                cert = open(cert, 'r').read().rstrip()
            else:
                cert = to_text(cert.rstrip())

        key = module.params.get('key')
        if key is not None:
            if os.path.isfile(key):
                key = open(key, 'r').read().rstrip()
            else:
                key = to_text(key.rstrip())

        cert_chain = module.params.get('cert_chain')
        if cert_chain is not None:
            if os.path.isfile(cert_chain):
                cert_chain = open(cert_chain, 'r').read()
            else:
                cert_chain = to_text(cert_chain.rstrip())
    else:
        cert=key=cert_chain=None
```

ProjectRice replied:
You'll also note I updated the else statement to correctly reflect the vars used later.
eg. cert_chain not chain.

s-hertel replied:
The changes we've made to this part of the module is responsible for the conflicts with https://github.com/ansible/ansible/commit/62fa2019c6846def66912b94ba1b5ed0c40525d and will be removed when rebasing. Should I go ahead and rebase so just changes for arn will remain?

ProjectRice replied:
Hi @s-hertel.

I'm happy to get this one merged so I can rebase and get some movement on https://github.com/ansible/ansible/pull/20622.  Ultimately this was @wrouesnel's PR though.

I note https://github.com/ansible/ansible/commit/62fa2019c6846def66912b94ba1b5ed0c40525d still has some of the issues rasied here.  [Line 270](https://github.com/ansible/ansible/blob/62fa2019c6846def66912b94ba1b5ed0c40525d0/lib/ansible/modules/cloud/amazon/iam_cert.py#L270) still defines `chain` which is never used and should be `cert_chain`. [Line 260](https://github.com/ansible/ansible/blob/62fa2019c6846def66912b94ba1b5ed0c40525d0/lib/ansible/modules/cloud/amazon/iam_cert.py#L260) could be removed if this was fixed.  I guess I'll fix those in https://github.com/ansible/ansible/pull/20622


Unfortunately there seems to be a bit of inconsistency in reviews.  On https://github.com/ansible/ansible/pull/20622 I was asked to remove empty `aliases: []` and `required: false` lines as part of my commit, yet there is no mention of it on this PR and on https://github.com/ansible/ansible/commit/62fa2019c6846def66912b94ba1b5ed0c40525d only the empty `aliases: []` entries have been removed.

wrouesnel replied:
I'll rebase this up to current devel again. Unfortunately since I'm not actively using AWS at the moment, chasing these PRs has been a challenge.

ProjectRice reviewed (changes requested):
See comments.

wrouesnel commented:
Okay so I was rebasing, and looking at this module now it looks like the net effect of this PR has already been merged?
```
EXAMPLES = '''
# Basic server certificate upload from local file
- iam_cert:
    name: very_ssl
    state: present
    cert: "{{ lookup('file', 'path/to/cert') }}"
    key: "{{ lookup('file', 'path/to/key') }}"
    cert_chain: "{{ lookup('file', 'path/to/certchain') }}"

# Server certificate upload using key string
- iam_cert:
    name: very_ssl
    state: present
    path: "/a/cert/path/"
    cert: body_of_somecert
    key: vault_body_of_privcertkey
    cert_chain: body_of_myverytrustedchain
'''
```
Would seem to accomplish the same thing as this PR? In which case this can be closed since the functionality has been implemented by 62fa2019c6846def66912b94ba1b5ed0c40525d0

ProjectRice commented:
Rebase this as you wish and I'll correct any issues I feel still exist in my outstanding PR.

Yes.  Seems only the ARN stuff is still required from your PR

wrouesnel committed:
iam_cert.py Fix duplicate certificate detection with included chains.

The iam_cert module would fail to detect certificates as duplicates
if the certificate body included the authority chain directly.

This commit fixes the problem by checking if a given certificate
matches the start of the data returned by AWS, since in all cases
where they would match the certificate will come first.

wrouesnel committed:
iam_cert.py Return certificate ARN in all success cases.

When uploading certificates or interacting with IAM, the certificate ARN
is needed for other operations with AWS such as provisioning elastic load
balancers.

This commit returns the certificate ARN in all success cases, which allows
it to be used to idempotently provision other Amazon services depending on
it (ELBs being an immediate example).

wrouesnel commented:
@ryansb Could you take a look? Your original comment was rendered moot by #24206 implementing that functionality, so this is just the ARN and duplicate certificate detection now.

@ProjectRice Rebased (and separated into 2 logical commits for the two changes, with detailed explanations).

ProjectRice commented:
Thank you all..  I will look at updating https://github.com/ansible/ansible/pull/20622 soon to reflect updates here.
